examples/simple-inference/simple-inference.cpp

```diff
@@ -0,0 +1,401 @@
+    }
+    console::set_display(console::reset);
+    return true;
+}
+
+
+int main(int argc, char ** argv) {
+    gpt_params params;
+
+    if (gpt_params_parse(argc, argv, params) == false) {
+        return 1;
+    }
+
+    if (!check_unsupported(&params)) {
+        return 1;
+    }
+
+#ifndef LOG_DISABLE_LOGS
+    log_set_target(log_filename_generator("simple-inference", "log"));
+    LOG_TEE("Log start\n");
+    log_dump_cmdline(argc,argv);
+#endif // LOG_DISABLE_LOGS
+
+    llama_context *     ctx = NULL;
+    llama_model *     model = NULL;
+    llama_grammar * grammar = NULL;
+    std::vector<llama_token> prompt_tokens;
+
+    if (!initialize(&ctx, &model, params, prompt_tokens, &grammar)) {
+        return 1;
+    }
+
+    const int n_ctx   = llama_n_ctx(ctx);
+    int n_remain      = params.n_predict;
+    std::vector<int>   input_tokens;
+
+    {
+        LOG("warming up the model with an empty run\n");
+
+        const std::vector<llama_token> tmp = { llama_token_bos(ctx), };
+        llama_eval(ctx, tmp.data(), tmp.size(), 0, params.n_threads);
+        llama_reset_timings(ctx);
+    }
+
+    if (!feed_prompt(ctx, &params, prompt_tokens.data(), prompt_tokens.size(), 0)) {
+        return 1;
+    }
+
+    if (n_remain < 0 || n_remain + int(prompt_tokens.size()) > n_ctx) {
+        n_remain = n_ctx - prompt_tokens.size();
+    }
+
+    std::vector<llama_token> last_tokens = prompt_tokens;
+    last_tokens.reserve(params.n_ctx);
+
+    std::vector<llama_token_data> candidates;
+    candidates.reserve(llama_n_vocab(ctx));
+
+    while (n_remain > 0 && interrupted.load() == false) {
+        const llama_token id = llama_sample_token(ctx, NULL, grammar, params, last_tokens, candidates);
+
+        last_tokens.push_back(id);
+        --n_remain;
+
+        LOG("n_remain: %d\n", n_remain);
+
+        // end of text token
+        if (id == llama_token_eos(ctx)) {
+            LOG_TEE(" [end of text]\n");
+            break;
+        }
+
+        const std::string token_str = llama_token_to_piece(ctx, id);
+        fputs(token_str.c_str(), stdout);
+        fflush(stdout);
+
+        // predict
+        if (n_remain > 0 && llama_eval(ctx, &id, 1, last_tokens.size(), params.n_threads)) {
+            LOG_TEE("%s : failed to eval\n", __func__);
+            return 1;
+        }
+    }
+
+    std::vector<int>   output_tokens;
+    std::ostringstream output_ss;
+    const size_t prompt_size = prompt_tokens.size();
+
+    for (size_t i = 0; i < last_tokens.size(); i++) {
+        const std::string token_str = llama_token_to_piece(ctx, last_tokens[i]);
+        if (i >= prompt_size) {
+            output_ss << token_str;
+            output_tokens.push_back(last_tokens[i]);
+        }
+
+    }
+
+    console::cleanup();
+    printf("\n");
+
+    llama_print_timings(ctx);
+    write_logfile(ctx, params, model, prompt_tokens, output_ss.str(), output_tokens);
+
+    llama_free(ctx);
+    llama_free_model(model);
+
+    if (grammar != NULL) {
```

Comment:
This null check is redundant - llama_grammar_free just calls `delete`, which is explicitly documented to accept NULL.

Reply:
I'm not sure the caller can assume that though? Ideally if that function got changed to do more stuff in the future, it would check for `NULL` though. I'm inclined to leave this as is (was also copied from `main`) just because it doesn't hurt anything and is more future proof.

Reply:
Well, if I were in charge of the API I would guarantee that all `*_free()` functions will accept NULL, just like `free()`, `delete`, and functions like GLib's `g_free()` do, simply because this pattern of freeing an optional object at the end of a function is so common.